packages/core/useContainerQuery/index.ts

```diff
@@ -0,0 +1,92 @@
+import { computed, ref } from 'vue-demi'
+import type { MaybeElementRef } from '@vueuse/core'
+import { useResizeObserver } from '../useResizeObserver'
+
+export interface Breakpoint {
+  /**
+   *
+   * minimum value of breakpoint (in pixels)
+   *
+   */
+  min?: number
+  /**
+   *
+   * maximum value of breakpoint (in pixels) *optional
+   *
+   */
+  max?: number
+}
+
+export type QueryBreakpoints = Record<string, Breakpoint>
+
+export interface UseContainerQueryOptions {
+  /**
+   *
+   * Target Element for the container query
+   *
+   */
+  el: MaybeElementRef
+  /**
+   *
+   * Explicit key/value map of the predefined breakpoints
+   * you wish to use.
+   *
+   * N.B. Each value should have a min and max range.
+   *
+   * (e.g., { sm: { min: 0, max: 300}, md: { min: 301, max: 600 }} )
+   *
+   */
+  breakpoints?: QueryBreakpoints
+}
+
+/**
+ *
+ * useContainerQuery()
+ *
+ * Element based media query to apply styles based on specified container elements
+ *
+ * @param options
+ * @see https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Container_Queries
+ * @see https://caniuse.com/css-container-queries
+ * @returns the active breakpoint from your specific list of breakpoints e.g,. sm
+ */
+export function useContainerQuery(options: UseContainerQueryOptions) {
+  // Set the default element and breakpoints:
+  const {
+    el,
+    breakpoints = {
+      sm: {
+        max: 480,
+      },
+      md: {
+        min: 481,
+        max: 768,
+      },
+      lg: {
+        min: 769,
+        max: 1024,
+      },
+      xl: {
+        min: 1025,
+        max: 1200,
+      },
+      xxl: {
+        min: 1201,
+      },
+    },
+  } = options
+
+  const width = ref(0)
+
+  useResizeObserver(el, ([entry]) => width.value = Math.round(entry.contentRect.width))
+
+  const activeBreakpoint = computed(() => {
+    for (const [key, { min = 0, max }] of Object.entries(breakpoints))
+      if (width.value >= min && (max === undefined || width.value <= max)) return key
+  })
+
+  return {
+    activeBreakpoint,
+    width,
+  }
+}
```

Comment:
Sorry. I accidentally pushed directly to your repository, I think this looks better. what do you think?

Reply:
Hi @okxiaoliang4 I believe the computed prop for activeBreakpoint is going to be slightly less performant this way because we don't want to re-computed on every change of width ... just when it changes boundary.

Reply:
https://github.com/vueuse/vueuse/pull/1297/commits/caff3bd3c0236685ab743999b44650a38fd43f9b#diff-d745801b80f40092b894f70c524ddaa9b8e7338eb3bc921d5c7a80f5f4565b49R120-R124

but you are watching `width`, also re-computed every time `width` changes

```typescript
watch(width, (newWidth: number, prevWidth: number) => {
  if (newWidth === prevWidth) return

  handleElementResize()
})
```

Reply:
Ah no actually with the watcher removed, it's slightly better! It seems less memory intensive. Nice! :)

Reply:
and you can try this. only when boundary changes will the `ref` update be triggered

```typescript
watch(activeBreakpoint, value => console.log(value))
```

![image](https://user-images.githubusercontent.com/22525904/154912051-709c2359-1c3b-488b-baaa-2e2f2058d301.png)

Reply:
and use computed to change breakpoints to `MaybeRef<QueryBreakpoints>`. can also trigger recalculation when `options.breakPoints` changes

Reply:
Yeh agree, it's more performant and should re-computed on breakpoint change.